File: scratch/test_eval.py
import json
import subprocess

with open('robo-delivery/index.html', 'r', encoding='utf-8') as f:
    content = f.read()

s_start = content.find('<script>') + len('<script>')
s_end = content.find('</script>')
js_code = content[s_start:s_end]

eval_js = f"""
const vm = require('vm');
const mockElement = {{
    getContext: () => ({{
        save: () => {{}}, restore: () => {{}}, translate: () => {{}}, scale: () => {{}},
        rotate: () => {{}}, beginPath: () => {{}}, moveTo: () => {{}}, lineTo: () => {{}},
        stroke: () => {{}}, fill: () => {{}}, drawImage: () => {{}}, arc: () => {{}},
        clearRect: () => {{}}, fillRect: () => {{}}, strokeText: () => {{}}, fillText: () => {{}}
    }}),
    style: {{}},
    addEventListener: () => {{}}
}};

const sandbox = {{
    console: console,
    document: {{
        getElementById: () => mockElement,
        createElement: () => mockElement,
        body: mockElement,
        addEventListener: () => {{}}
    }},
    window: {{
        innerWidth: 800,
        innerHeight: 600,
        addEventListener: () => {{}},
        removeEventListener: () => {{}}
    }},
    Image: class {{
        constructor() {{ setTimeout(() => {{ if (this.onload) this.onload(); }}, 10); }}
    }},
    AudioContext: class {{}},
    setTimeout: setTimeout,
    clearTimeout: clearTimeout,
    setInterval: setInterval,
    clearInterval: clearInterval,
    requestAnimationFrame: (cb) => setTimeout(cb, 16)
}};
sandbox.window.window = sandbox.window;
sandbox.window.document = sandbox.document;

const code = {json.dumps(js_code)};

try {{
    vm.createContext(sandbox);
    vm.runInContext(code, sandbox);
    console.log("SUCCESS_EVAL: No syntax or initialization errors found!");
}} catch(e) {{
    console.error("FAIL_EVAL:", e.message);
    console.error(e.stack);
}}
"""

with open('scratch/run_eval.js', 'w', encoding='utf-8') as f:
    f.write(eval_js)

res = subprocess.run(['node', 'scratch/run_eval.js'], capture_output=True, text=True)
print('STDOUT:', res.stdout)
print('STDERR:', res.stderr)

File: scratch/test_eval.js
const fs = require("fs");
let html = fs.readFileSync("robo-delivery/index.html", "utf-8");

global.window = global;
global.window.addEventListener = () => {};
global.window.removeEventListener = () => {};
global.innerWidth = 800;
global.innerHeight = 600;
global.document = {
  getElementById: (id) => ({
    getContext: () => ({
      save: () => {}, restore: () => {}, fillRect: () => {}, drawImage: () => {},
      beginPath: () => {}, arc: () => {}, fill: () => {}, stroke: () => {},
      moveTo: () => {}, lineTo: () => {}, setLineDash: () => {}, createLinearGradient: () => ({ addColorStop: () => {} }),
      createRadialGradient: () => ({ addColorStop: () => {} }), font: '', textAlign: '', textBaseline: '', fillText: () => {},
      classList: { add: () => {}, remove: () => {} }, style: {}, roundRect: () => {}
    }),
    width: 800, height: 600, addEventListener: () => {}, classList: { add: () => {}, remove: () => {} }, style: {},
    getBoundingClientRect: () => ({ left: 0, top: 0, width: 800, height: 600 }),
    innerText: ''
  }),
  querySelector: () => null, querySelectorAll: () => [], addEventListener: () => {}
};
global.Image = class { constructor() { this.width = 1024; this.height = 1024; } };
global.requestAnimationFrame = (fn) => {};

// Insert global state declarations at start of script
const globals = `
    const canvas = document.getElementById('gameCanvas');
    const ctx = canvas ? canvas.getContext('2d') : null;
    let scaleFactor = 1.0;
    let hoveredNodeId = null;
    let isGameRunning = false;
    let timerInterval = null;
    let currentLevelIndex = 0;
    let currentLevelData = null;

    const robotState = {
      currentNodeId: 0,
      prevNodeId: 0,
      targetNodeId: null,
      isMoving: false,
      moveProgress: 0,
      carryingFood: null,
      bumpAnim: 0,
      x: 0,
      y: 0,
      z: 0
    };
`;

html = html.replace('<script>', '<script>\n' + globals);

const scriptMatch = html.match(/<script[\s\S]*?>([\s\S]*?)<\/script>/i);
if (scriptMatch) {
  try {
    eval(scriptMatch[1]);
    console.log("SUCCESS! Script executed completely without ANY runtime error!");
  } catch(e) {
    console.error("ERROR:", e.message, e.stack);
  }
}
